src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectKey.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+/// <summary>
+/// A very light wrapper around a file path, used to ensure consistency across the code base for what constitutes the unique
+/// identifier for a project.
+/// </summary>
+internal sealed class ProjectKey
+{
+    // ProjectKey represents the full path to the project file name. All creation logic is here in one place
+    // to ensure this is consistent.
+    public static ProjectKey From(HostProject hostProject) => new(hostProject.FilePath);
+    public static ProjectKey From(IProjectSnapshot project) => new(project.FilePath);
+    public static ProjectKey? From(Project project) => project.FilePath is null ? null : new(project.FilePath);
+    public static ProjectKey From(string projectFilePath)
```

Comment:
The aim is, by the end of this effort, that this method is gone completely. Therefore, people would have to create a project (one of the three forms above) correctly, before they can produce a key, therefore we'll be able to trust the keys.